### ISSUE TITLE
Incorrect placement of Colorbar ticks using LogNorm

### ISSUE THREAD

**Issue by rayosborn** [matplotlib/matplotlib#12155]
### Bug report

**Bug summary**

After upgrading to Matplotlib v3.0.0, the tick marks on the colorbar are no longer placed correctly when using a matplotlib.colors.LogNorm normalization. 

**Code for reproduction**

The code is embedded in a class defined within the NeXpy application, but uses standard Matplotlib function calls. Here are the relevant extracted lines:

```python
from matplotlib.colors import LogNorm, Normalize, SymLogNorm
from matplotlib.ticker import AutoLocator, LogLocator, ScalarFormatter
from matplotlib.ticker import LogFormatterSciNotation as LogFormatter
        self.norm = LogNorm(self.vaxis.lo, self.vaxis.hi)
        self.locator = LogLocator()
        self.formatter = LogFormatter()
        self.image = ax.pcolormesh(x, y, self.v, cmap=self.cmap, **opts)
        self.image.set_norm(self.norm)
        self.colorbar = self.figure.colorbar(self.image, ax=ax, norm=self.norm)
        self.colorbar.locator = self.locator
        self.colorbar.formatter = self.formatter
        self.colorbar.update_normal(self.image)
```

**Actual outcome**

In Matplotlib v3.0.0, I get:
![matplotlib_3_0_0](https://user-images.githubusercontent.com/1193959/45703125-0cf96100-bb39-11e8-81e8-70de6cb87b3e.png)

**Expected outcome**

In Matplotlib v2.2.3, I get:
![matplotlib_2_2_3](https://user-images.githubusercontent.com/1193959/45703101-fe12ae80-bb38-11e8-9462-8abe64722570.png)

**Matplotlib version**
  * Operating system: Mac OS 10.12.6
  * Matplotlib version: 3.0.0
  * Matplotlib backend: Qt5Agg
  * Python version: 3.6.5

I got the same result when using conda to install v3.0.0 or pip to install the latest development version.

**Comment by tacaswell** [comment]
That looks like the 'log' scale did not get propogated to the color bar axex correctly.

If you do `self.colorbar.set_yscale('log')` does it look right?

**Comment by ImportanceOfBeingErnest** [comment]
I do have problems reproducing this. I ran the following in the current development version

```
import matplotlib
print(matplotlib.__version__)  # 3.0.0rc1.post326+g521b60a77
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
from matplotlib.ticker import LogLocator, LogFormatter
import numpy as np

x,y = np.ogrid[-4:4:31j,-4:4:31j]
z = np.exp(-x**2-y**2)

fig, ax = plt.subplots()

norm = mcolors.LogNorm(z.min(), z.max())

im = ax.pcolormesh(z, norm=norm)
im.set_norm(norm)

locator = LogLocator()
formatter = LogFormatter()

cbar = fig.colorbar(im, ax=ax,norm=norm)
cbar.locator = locator
cbar.formatter = formatter
cbar.update_normal(im)

plt.show()
```

and this is the result which looks expected:

![image](https://user-images.githubusercontent.com/23121882/45709681-d77a6500-bb84-11e8-8f28-7ae15bd6168a.png)

You may notice that the ticklabels look differently compared to the example which makes me think that the code shown in the original post is not complete. A [Short, Self Contained, Correct (Compilable), Example](http://sscce.org) would come handy here, else it's probably hard to ever find out what's going on.

**Comment by rayosborn** [comment]
I will see if I can reproduce it outside the NeXpy context. The reason I posted is that this specific code has been untouched since Matplotlib v1.5, and it has worked without problem up until v2.2.3. I was hoping someone could point to the changes in the Matplotlib API that are likely to be relevant.

**Comment by jklymak** [comment]
Coorbar handling definitely changed:

**Comment by rayosborn** [comment]
I've managed to produce a simple example demonstrating the problem. It turns out that the issue shows up if you make a plot with a linear color scale first and then transform it to a log scale.

```
import matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
from matplotlib.ticker import LogLocator, LogFormatter
import numpy as np

x,y = np.ogrid[-4:4:31j,-4:4:31j]
z = 120000*np.exp(-x**2-y**2)

fig, ax = plt.subplots()

im = ax.imshow(z)
cbar = fig.colorbar(im)

norm = LogNorm(z.min(), z.max())
im.set_norm(norm)
cbar.set_norm(norm)
cbar.locator = LogLocator()
cbar.formatter = LogFormatter()
cbar.update_normal(im)

plt.show()
```
![matplotlib-color-bar-issue](https://user-images.githubusercontent.com/1193959/45715362-b5b7b880-bb59-11e8-8ec1-50dfca17ebd1.png)

**Comment by ImportanceOfBeingErnest** [comment]
An immediate solution is to use  

     cbar.update_bruteforce(im)

instead.

**Comment by rayosborn** [comment]
Thanks for giving such quick feedback. NeXpy has a checkbox for switching between linear and log scales, which we use all the time when looking for features on different intensity scales. It looks like replacing `update_normal` with `update_bruteforce` fixes my problem in NeXpy - if I use `update_normal` and `config_axis`, then the height of my color bar changes when I change the intensity scale. Is this what the `shrink` option is for?

**Comment by jklymak** [comment]
What do you do to change the intensity scale?  If I do the following, it all seems to work fine, but maybe I'm misunderstanding.  You shouldn't have to play around w/ shrink after you have initially made the colorbar.

```python
norm = LogNorm(z.min(), z.max())
im.set_norm(norm)
cbar.set_norm(norm)
cbar.update_normal(im)
cbar.config_axis()

norm = LogNorm(z.min(), z.max()*10)
im.set_norm(norm)
cbar.set_norm(norm)
cbar.update_normal(im)
cbar.config_axis()

```

**Comment by ImportanceOfBeingErnest** [comment]
To reproduce the shrinking: 

```
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

x,y = np.ogrid[-4:4:31j,-4:4:31j]
z = 120000*np.exp(-x**2-y**2)

fig, ax = plt.subplots()

im = ax.imshow(z)
cbar = fig.colorbar(im)

# nopw change the data and scale
z*=1000
im.set_data(z)

norm = LogNorm(z.min(), z.max())
im.set_norm(norm)
cbar.set_norm(norm)

cbar.update_normal(im)
cbar.config_axis()

plt.show()
```

![image](https://user-images.githubusercontent.com/23121882/45718959-1f0dea80-bb9f-11e8-8f9c-1ebd1f781e28.png)

**Comment by jklymak** [comment]
Huh.  That’s a very strange bug.  I’ll check it out soon.

**Comment by jklymak** [comment]
```python
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

x,y = np.ogrid[-4:4:31j,-4:4:31j]
z = 120000*np.exp(-x**2-y**2)

fig, ax = plt.subplots()

im = ax.imshow(z)
cbar = fig.colorbar(im)

norm = LogNorm(z.min(), z.max())
im.set_norm(norm)
cbar.set_norm(norm)
cbar.update_normal(im)
cbar.config_axis()

# display in here, and then user changes zmax * and* zmin...

norm = LogNorm(z.min() * 1000, z.max() * 1000)
im.set_norm(norm)
cbar.set_norm(norm)
cbar.update_normal(im)
cbar.config_axis()

plt.show()
```

**Comment by jklymak** [comment]
Fix in #12159 now fixes the above as well.  Needs a test I guess...

**Comment by jklymak** [comment]
I think #12159 fixes this, but I guess I think if a GUI is going to change so much of the colorbar, it should indeed call `cbar.update_bruteforce(im)`.  I'm not thrilled with the name of this function, but it does what the OP wanted, which is to remove the old colorbar and make a new colorbar based on `im`.  The info about locators etc is lost, but presumably the user has kept track of those.  

That it used to work before is because we drew colorbars ticks by hand, basically.  That it doesn't work now is that colorbars behave more like a normal axes and let the normal tick creation do its thing.

Note that `self.locator = LogLocator()` will tend to add extra ticks now if `extend='both'`.  We added `colorbar._ColorbarLogLocator` to trim the extra ticks.  We are trying to add that functionality to all the locators soon, but we are not there yet.
